File: backend/src/services/dashboard.service.js
const pool = require('../config/database');

const STATUS_LABELS = {
  active: 'Đang học',
  paused: 'Tạm nghỉ',
  graduated: 'Đã tốt nghiệp',
  dropped: 'Đã nghỉ',
};

function hasRole(user, role) {
  return (user?.roles || []).includes(role);
}

class DashboardService {
  async getStats(tenantId, user) {
    if (hasRole(user, 'admin') || hasRole(user, 'staff')) {
      return this._getAdminStaffStats(tenantId);
    }
    if (hasRole(user, 'teacher')) {
      return this._getTeacherStats(tenantId, user.id);
    }
    return this._getStudentStats(tenantId, user.id);
  }

  async _getAdminStaffStats(tenantId) {
    const [
      students,
      classes,
      teachers,
      todaySessions,
      pendingHomework,
      pendingMockExams,
      weeklyAttendance,
      studentStatusDist,
      topClasses,
      newStudentsByMonth,
      classTypeDist,
    ] = await Promise.all([
      pool.query(
        `SELECT COUNT(*) FROM users u JOIN roles r ON r.user_id=u.id
         WHERE u.tenant_id=$1 AND r.role_type='student' AND u.is_active=TRUE`,
        [tenantId]
      ),
      pool.query(`SELECT COUNT(*) FROM classes WHERE tenant_id=$1 AND status='active'`, [tenantId]),
      pool.query(
        `SELECT COUNT(DISTINCT u.id)
         FROM users u JOIN roles r ON r.user_id = u.id AND r.tenant_id = u.tenant_id
         WHERE u.tenant_id=$1 AND r.role_type='teacher' AND u.is_active=TRUE`,
        [tenantId]
      ),
      pool.query(
        `SELECT COUNT(*) FROM schedules WHERE tenant_id=$1 AND session_date=CURRENT_DATE AND status='scheduled'`,
        [tenantId]
      ),
      pool.query(
        `SELECT COUNT(*) FROM homework_submissions hs
         JOIN homework_assignment_students has ON has.id = hs.assignment_student_id
         JOIN homework_assignments ha ON ha.id = has.assignment_id
         WHERE ha.tenant_id=$1 AND hs.status='submitted'`,
        [tenantId]
      ).catch(() => ({ rows: [{ count: 0 }] })),
      pool.query(
        `SELECT COUNT(*) FROM mock_exam_students mes
         JOIN mock_exams me ON me.id = mes.mock_exam_id
         WHERE me.tenant_id=$1 AND mes.status='submitted'`,
        [tenantId]
      ).catch(() => ({ rows: [{ count: 0 }] })),
      pool.query(
        `SELECT s.session_date,
                COUNT(*) FILTER (WHERE a.status = 'present') as present_count,
                COUNT(*) as total_count
         FROM attendance a
         JOIN schedules s ON s.id = a.schedule_id
         WHERE a.tenant_id=$1 AND s.session_date >= CURRENT_DATE - INTERVAL '6 days' AND s.session_date <= CURRENT_DATE
         GROUP BY s.session_date
         ORDER BY s.session_date`,
        [tenantId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT sp.study_status, COUNT(*) as count
         FROM student_profiles sp
         WHERE sp.tenant_id=$1
         GROUP BY sp.study_status`,
        [tenantId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT c.name, COUNT(cs.student_id) as student_count
         FROM classes c
         JOIN class_students cs ON cs.class_id=c.id AND cs.status='active'
         WHERE c.tenant_id=$1 AND c.status='active'
         GROUP BY c.id, c.name
         ORDER BY student_count DESC
         LIMIT 5`,
        [tenantId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT to_char(date_trunc('month', enrollment_date), 'YYYY-MM') AS month, COUNT(*) AS count
         FROM student_profiles
         WHERE tenant_id=$1 AND enrollment_date >= CURRENT_DATE - INTERVAL '6 months'
         GROUP BY 1 ORDER BY 1`,
        [tenantId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT class_type, COUNT(*) AS count FROM classes
         WHERE tenant_id=$1 AND status='active' GROUP BY class_type`,
        [tenantId]
      ).catch(() => ({ rows: [] })),
    ]);

    return {
      role: 'admin',
      students: parseInt(students.rows[0].count, 10),
      activeClasses: parseInt(classes.rows[0].count, 10),
      teachers: parseInt(teachers.rows[0].count, 10),
      todaySessions: parseInt(todaySessions.rows[0].count, 10),
      pendingHomeworkGrading: parseInt(pendingHomework.rows[0]?.count || 0, 10),
      pendingMockExamGrading: parseInt(pendingMockExams.rows[0]?.count || 0, 10),
      weeklyAttendance: weeklyAttendance.rows.map((row) => ({
        date: row.session_date,
        rate: Number(row.total_count) > 0 ? Math.round((Number(row.present_count) / Number(row.total_count)) * 100) : 0,
        present: Number(row.present_count),
        total: Number(row.total_count),
      })),
      studentStatusDistribution: studentStatusDist.rows.map((row) => ({
        status: row.study_status,
        label: STATUS_LABELS[row.study_status] || row.study_status,
        count: Number(row.count),
      })),
      topClasses: topClasses.rows.map((row) => ({ name: row.name, count: Number(row.student_count) })),
      newStudentsByMonth: newStudentsByMonth.rows.map((row) => ({ month: row.month, count: Number(row.count) })),
      classTypeDistribution: classTypeDist.rows.map((row) => ({ type: row.class_type, count: Number(row.count) })),
    };
  }

  async _getTeacherStats(tenantId, teacherId) {
    const [
      myClasses,
      todaySessions,
      myStudents,
      pendingHomework,
      pendingMockExams,
      weekSchedule,
      recentHomework,
      teacherWeeklyAttendance,
      gradingProgress,
    ] = await Promise.all([
      pool.query(
        `SELECT COUNT(DISTINCT ct.class_id) FROM class_teachers ct
         JOIN classes c ON c.id = ct.class_id
         WHERE ct.tenant_id=$1 AND ct.teacher_id=$2 AND c.status='active'`,
        [tenantId, teacherId]
      ),
      pool.query(
        `SELECT COUNT(*) FROM schedules
         WHERE tenant_id=$1 AND teacher_id=$2 AND session_date=CURRENT_DATE AND status='scheduled'`,
        [tenantId, teacherId]
      ),
      pool.query(
        `SELECT COUNT(DISTINCT cs.student_id)
         FROM class_teachers ct
         JOIN class_students cs ON cs.class_id = ct.class_id AND cs.tenant_id = ct.tenant_id AND cs.status='active'
         WHERE ct.tenant_id=$1 AND ct.teacher_id=$2`,
        [tenantId, teacherId]
      ),
      pool.query(
        `SELECT COUNT(*) FROM homework_submissions hs
         JOIN homework_assignment_students has ON has.id = hs.assignment_student_id
         JOIN homework_assignments ha ON ha.id = has.assignment_id
         WHERE ha.tenant_id=$1 AND hs.status='submitted' AND ha.created_by=$2`,
        [tenantId, teacherId]
      ).catch(() => ({ rows: [{ count: 0 }] })),
      pool.query(
        `SELECT COUNT(*) FROM mock_exam_students mes
         JOIN mock_exams me ON me.id = mes.mock_exam_id
         JOIN class_teachers ct ON ct.class_id = me.class_id AND ct.tenant_id = me.tenant_id
         WHERE me.tenant_id=$1 AND mes.status='submitted' AND ct.teacher_id=$2`,
        [tenantId, teacherId]
      ).catch(() => ({ rows: [{ count: 0 }] })),
      pool.query(
        `SELECT s.session_date, s.start_time, s.end_time, c.name AS class_name, r.name AS room_name
         FROM schedules s
         JOIN classes c ON c.id = s.class_id
         LEFT JOIN rooms r ON r.id = s.room_id
         WHERE s.tenant_id=$1 AND s.teacher_id=$2 AND s.status='scheduled'
           AND s.session_date BETWEEN CURRENT_DATE AND CURRENT_DATE + INTERVAL '6 days'
         ORDER BY s.session_date, s.start_time
         LIMIT 8`,
        [tenantId, teacherId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT ha.id, ha.title,
                (SELECT string_agg(c.name, ', ') FROM homework_assignment_classes hac JOIN classes c ON c.id=hac.class_id WHERE hac.assignment_id=ha.id) AS class_names,
                (SELECT COUNT(*) FROM homework_submissions hs JOIN homework_assignment_students has ON has.id=hs.assignment_student_id WHERE has.assignment_id=ha.id AND hs.status='submitted') AS pending_count
         FROM homework_assignments ha
         WHERE ha.tenant_id=$1 AND ha.created_by=$2 AND ha.status='active'
         ORDER BY ha.created_at DESC
         LIMIT 5`,
        [tenantId, teacherId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT s.session_date,
                COUNT(*) FILTER (WHERE a.status='present') AS present_count,
                COUNT(*) AS total_count
         FROM attendance a
         JOIN schedules s ON s.id = a.schedule_id
         WHERE a.tenant_id=$1 AND s.teacher_id=$2
           AND s.session_date >= CURRENT_DATE - INTERVAL '6 days' AND s.session_date <= CURRENT_DATE
         GROUP BY s.session_date ORDER BY s.session_date`,
        [tenantId, teacherId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT
           COUNT(*) FILTER (WHERE hs.status='graded') AS graded_count,
           COUNT(*) FILTER (WHERE hs.status='submitted') AS pending_count
         FROM homework_submissions hs
         JOIN homework_assignment_students has ON has.id = hs.assignment_student_id
         JOIN homework_assignments ha ON ha.id = has.assignment_id
         WHERE ha.tenant_id=$1 AND ha.created_by=$2`,
        [tenantId, teacherId]
      ).catch(() => ({ rows: [{ graded_count: 0, pending_count: 0 }] })),
    ]);

    return {
      role: 'teacher',
      activeClasses: parseInt(myClasses.rows[0].count, 10),
      todaySessions: parseInt(todaySessions.rows[0].count, 10),
      students: parseInt(myStudents.rows[0].count, 10),
      pendingHomeworkGrading: parseInt(pendingHomework.rows[0]?.count || 0, 10),
      pendingMockExamGrading: parseInt(pendingMockExams.rows[0]?.count || 0, 10),
      weekSchedule: weekSchedule.rows.map((row) => ({
        date: row.session_date,
        startTime: row.start_time,
        endTime: row.end_time,
        className: row.class_name,
        roomName: row.room_name,
      })),
      myHomeworks: recentHomework.rows.map((row) => ({
        id: row.id,
        title: row.title,
        classNames: row.class_names,
        pendingCount: Number(row.pending_count || 0),
      })),
      weeklyAttendance: teacherWeeklyAttendance.rows.map((row) => ({
        date: row.session_date,
        rate: Number(row.total_count) > 0 ? Math.round((Number(row.present_count) / Number(row.total_count)) * 100) : 0,
      })),
      gradingProgress: {
        graded: Number(gradingProgress.rows[0]?.graded_count || 0),
        pending: Number(gradingProgress.rows[0]?.pending_count || 0),
      },
    };
  }

  async _getStudentStats(tenantId, studentId) {
    const [
      myClasses,
      todaySessions,
      pendingHomework,
      upcomingMockExams,
      attendance,
      upcomingHomeworkList,
      recentScores,
      monthlyAttendance,
    ] = await Promise.all([
      pool.query(
        `SELECT COUNT(*) FROM class_students WHERE tenant_id=$1 AND student_id=$2 AND status='active'`,
        [tenantId, studentId]
      ),
      pool.query(
        `SELECT COUNT(*) FROM schedules s
         JOIN class_students cs ON cs.class_id = s.class_id AND cs.tenant_id = s.tenant_id
         WHERE s.tenant_id=$1 AND cs.student_id=$2 AND cs.status='active'
           AND s.session_date=CURRENT_DATE AND s.status='scheduled'`,
        [tenantId, studentId]
      ),
      pool.query(
        `SELECT COUNT(*) FROM homework_assignment_students has
         JOIN homework_assignments ha ON ha.id = has.assignment_id
         WHERE ha.tenant_id=$1 AND has.student_id=$2 AND ha.status='active'
           AND has.status IN ('assigned','in_progress','revision_required')`,
        [tenantId, studentId]
      ).catch(() => ({ rows: [{ count: 0 }] })),
      pool.query(
        `SELECT COUNT(*) FROM mock_exam_students mes
         JOIN mock_exams me ON me.id = mes.mock_exam_id
         WHERE me.tenant_id=$1 AND mes.student_id=$2 AND mes.status IN ('assigned','in_progress')`,
        [tenantId, studentId]
      ).catch(() => ({ rows: [{ count: 0 }] })),
      pool.query(
        `SELECT COUNT(*) FILTER (WHERE status='present') AS present_count, COUNT(*) AS total_count
         FROM attendance WHERE tenant_id=$1 AND student_id=$2`,
        [tenantId, studentId]
      ).catch(() => ({ rows: [{ present_count: 0, total_count: 0 }] })),
      pool.query(
        `SELECT ha.id, ha.title, ha.due_date,
                (SELECT string_agg(c.name, ', ') FROM homework_assignment_classes hac JOIN classes c ON c.id=hac.class_id WHERE hac.assignment_id=ha.id) AS class_names
         FROM homework_assignment_students has
         JOIN homework_assignments ha ON ha.id = has.assignment_id
         WHERE ha.tenant_id=$1 AND has.student_id=$2 AND ha.status='active'
           AND has.status IN ('assigned','in_progress','revision_required')
         ORDER BY ha.due_date ASC NULLS LAST
         LIMIT 5`,
        [tenantId, studentId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT ha.title, has.total_score, ha.total_score AS max_score, hs.graded_at
         FROM homework_submissions hs
         JOIN homework_assignment_students has ON has.id = hs.assignment_student_id
         JOIN homework_assignments ha ON ha.id = has.assignment_id
         WHERE ha.tenant_id=$1 AND has.student_id=$2 AND hs.status='graded'
         ORDER BY hs.graded_at DESC LIMIT 5`,
        [tenantId, studentId]
      ).catch(() => ({ rows: [] })),
      pool.query(
        `SELECT to_char(date_trunc('month', s.session_date), 'YYYY-MM') AS month,
                COUNT(*) FILTER (WHERE a.status='present') AS present_count,
                COUNT(*) AS total_count
         FROM attendance a
         JOIN schedules s ON s.id = a.schedule_id
         WHERE a.tenant_id=$1 AND a.student_id=$2 AND s.session_date >= CURRENT_DATE - INTERVAL '6 months'
         GROUP BY 1 ORDER BY 1`,
        [tenantId, studentId]
      ).catch(() => ({ rows: [] })),
    ]);

    const presentCount = Number(attendance.rows[0]?.present_count || 0);
    const totalCount = Number(attendance.rows[0]?.total_count || 0);

    return {
      role: 'student',
      activeClasses: parseInt(myClasses.rows[0].count, 10),
      todaySessions: parseInt(todaySessions.rows[0].count, 10),
      pendingHomeworkCount: parseInt(pendingHomework.rows[0]?.count || 0, 10),
      upcomingMockExamCount: parseInt(upcomingMockExams.rows[0]?.count || 0, 10),
      attendanceRate: totalCount > 0 ? Math.round((presentCount / totalCount) * 100) : null,
      attendancePresent: presentCount,
      attendanceTotal: totalCount,
      upcomingHomework: upcomingHomeworkList.rows.map((row) => ({
        id: row.id,
        title: row.title,
        dueDate: row.due_date,
        classNames: row.class_names,
      })),
      recentScores: recentScores.rows.map((row) => ({
        title: row.title,
        score: Number(row.total_score || 0),
        maxScore: Number(row.max_score || 100),
      })),
      monthlyAttendance: monthlyAttendance.rows.map((row) => ({
        month: row.month,
        rate: Number(row.total_count) > 0 ? Math.round((Number(row.present_count) / Number(row.total_count)) * 100) : 0,
      })),
    };
  }
}

module.exports = new DashboardService();

File: backend/Services/DashboardService.cs
using LanguageCenter.Api.Infrastructure;

namespace LanguageCenter.Api.Services;

public sealed class DashboardService
{
    private readonly DbService _db;
    public DashboardService(DbService db) => _db = db;

    public async Task<object> GetStatsAsync(int tenantId, int? month = null, int? year = null)
    {
        var now = DateTime.Now;
        var m = month ?? now.Month;
        var y = year ?? now.Year;

        // ---- Thống kê tổng quan (flat, khớp Frontend) ----
        var students = await _db.ScalarAsync<long>("""
            SELECT COUNT(*) FROM users u
            JOIN roles r ON r.user_id=u.id AND r.tenant_id=u.tenant_id
            WHERE u.tenant_id=@tenantId AND r.role_type='student' AND u.is_active=TRUE
            """, new { tenantId });

        var teachers = await _db.ScalarAsync<long>("""
            SELECT COUNT(DISTINCT u.id) FROM users u
            JOIN roles r ON r.user_id=u.id AND r.tenant_id=u.tenant_id
            WHERE u.tenant_id=@tenantId AND r.role_type='teacher' AND u.is_active=TRUE
            """, new { tenantId });

        var activeClasses = await _db.ScalarAsync<long>("""
            SELECT COUNT(*) FROM classes WHERE tenant_id=@tenantId AND status='active'
            """, new { tenantId });

        var todaySessions = await _db.ScalarAsync<long>("""
            SELECT COUNT(*) FROM schedules
            WHERE tenant_id=@tenantId AND session_date=CURRENT_DATE AND status='scheduled'
            """, new { tenantId });

        // ---- Doanh thu theo phương thức và theo ngày ----
        var revenueByMethod = await _db.QueryAsync("""
            SELECT payment_method, COALESCE(SUM(amount),0) AS total, COUNT(*) AS count
            FROM fee_transactions
            WHERE tenant_id=@tenantId AND is_cancelled=FALSE
              AND EXTRACT(MONTH FROM paid_at)=@m AND EXTRACT(YEAR FROM paid_at)=@y
            GROUP BY payment_method
            """, new { tenantId, m, y });

        var revenueByDay = await _db.QueryAsync("""
            SELECT DATE(paid_at) AS day, SUM(amount) AS amount
            FROM fee_transactions
            WHERE tenant_id=@tenantId AND is_cancelled=FALSE
              AND EXTRACT(MONTH FROM paid_at)=@m AND EXTRACT(YEAR FROM paid_at)=@y
            GROUP BY DATE(paid_at) ORDER BY day
            """, new { tenantId, m, y });

        var totalRevenue = revenueByMethod.Sum(r => Convert.ToDouble(r.GetValueOrDefault("total") ?? 0));
        var totalTransactions = revenueByMethod.Sum(r => Convert.ToInt64(r.GetValueOrDefault("count") ?? 0));

        // ---- Chi phí ----
        var totalExpenses = await _db.ScalarAsync<double>("""
            SELECT COALESCE(SUM(amount),0) FROM expenses
            WHERE tenant_id=@tenantId AND is_active=TRUE
              AND EXTRACT(MONTH FROM expense_date)=@m AND EXTRACT(YEAR FROM expense_date)=@y
            """, new { tenantId, m, y });

        // ---- Hóa đơn chờ ----
        var pendingInvoice = await _db.QuerySingleAsync("""
            SELECT COUNT(*) AS count, COALESCE(SUM(fci.amount_due - fci.amount_paid),0) AS total
            FROM fee_collection_items fci
            JOIN fee_collections fc ON fc.id=fci.collection_id
            WHERE fci.tenant_id=@tenantId AND fci.status IN ('pending','partial') AND fc.status='active'
            """, new { tenantId });

        // ---- Công nợ theo nhóm tuổi ----
        var debtAgingRaw = await _db.QuerySingleAsync("""
            SELECT
              COUNT(DISTINCT CASE WHEN CURRENT_DATE - fc.due_date BETWEEN 1 AND 30  THEN fci.student_id END) AS days_1_30,
              COUNT(DISTINCT CASE WHEN CURRENT_DATE - fc.due_date BETWEEN 31 AND 60 THEN fci.student_id END) AS days_31_60,
              COUNT(DISTINCT CASE WHEN CURRENT_DATE - fc.due_date BETWEEN 61 AND 90 THEN fci.student_id END) AS days_61_90,
              COUNT(DISTINCT CASE WHEN CURRENT_DATE - fc.due_date > 90             THEN fci.student_id END) AS days_90_plus,
              COALESCE(SUM(CASE WHEN CURRENT_DATE - fc.due_date BETWEEN 1  AND 30 THEN fci.amount_due - fci.amount_paid ELSE 0 END),0) AS amt_1_30,
              COALESCE(SUM(CASE WHEN CURRENT_DATE - fc.due_date BETWEEN 31 AND 60 THEN fci.amount_due - fci.amount_paid ELSE 0 END),0) AS amt_31_60,
              COALESCE(SUM(CASE WHEN CURRENT_DATE - fc.due_date BETWEEN 61 AND 90 THEN fci.amount_due - fci.amount_paid ELSE 0 END),0) AS amt_61_90,
              COALESCE(SUM(CASE WHEN CURRENT_DATE - fc.due_date > 90              THEN fci.amount_due - fci.amount_paid ELSE 0 END),0) AS amt_90_plus,
              COUNT(DISTINCT fci.student_id) AS total_students_in_debt,
              COALESCE(SUM(fci.amount_due - fci.amount_paid),0) AS total_debt,
              COUNT(DISTINCT fci.id) AS total_overdue_items
            FROM fee_collection_items fci
            JOIN fee_collections fc ON fc.id = fci.collection_id
            WHERE fci.tenant_id=@tenantId AND fci.status IN ('pending','partial')
              AND fc.status='active' AND fc.due_date < CURRENT_DATE
            """, new { tenantId });

        var topDebtors = await _db.QueryAsync("""
            SELECT u.full_name, sp.student_code,
                   SUM(fci.amount_due - fci.amount_paid) AS debt_amount,
                   COUNT(*) AS overdue_collections,
                   MAX(CURRENT_DATE - fc.due_date) AS max_overdue_days
            FROM fee_collection_items fci
            JOIN fee_collections fc ON fc.id = fci.collection_id
            JOIN users u ON u.id = fci.student_id
            LEFT JOIN student_profiles sp ON sp.user_id = u.id AND sp.tenant_id = fci.tenant_id
            WHERE fci.tenant_id=@tenantId AND fci.status IN ('pending','partial')
              AND fc.status='active' AND fc.due_date < CURRENT_DATE
            GROUP BY u.full_name, sp.student_code
            ORDER BY debt_amount DESC LIMIT 10
            """, new { tenantId });

        // ---- Thống kê chi nhánh ----
        var branchStats = await _db.QueryAsync("""
            SELECT b.name AS branch_name, COUNT(DISTINCT cs.student_id) AS student_count
            FROM branches b
            JOIN classes c ON c.branch_id=b.id AND c.status='active'
            JOIN class_students cs ON cs.class_id=c.id AND cs.status='active'
            WHERE b.tenant_id=@tenantId
            GROUP BY b.id, b.name ORDER BY student_count DESC
            """, new { tenantId });

        // ---- Điểm danh tuần ----
        var weeklyAttendance = await _db.QueryAsync("""
            SELECT s.session_date,
                   COUNT(*) FILTER (WHERE a.status='present') AS present_count,
                   COUNT(*) AS total_count
            FROM attendance a
            JOIN schedules s ON s.id=a.schedule_id
            WHERE a.tenant_id=@tenantId
              AND s.session_date >= CURRENT_DATE - INTERVAL '6 days'
              AND s.session_date <= CURRENT_DATE
            GROUP BY s.session_date ORDER BY s.session_date
            """, new { tenantId });

        // ---- Phân phối trạng thái học viên ----
        var studentStatusDist = await _db.QueryAsync("""
            SELECT sp.study_status, COUNT(*) AS count
            FROM student_profiles sp
            WHERE sp.tenant_id=@tenantId
            GROUP BY sp.study_status
            """, new { tenantId });

        // ---- Top lớp học ----
        var topClasses = await _db.QueryAsync("""
            SELECT c.name, COUNT(cs.student_id) AS student_count
            FROM classes c
            JOIN class_students cs ON cs.class_id=c.id AND cs.status='active'
            WHERE c.tenant_id=@tenantId AND c.status='active'
            GROUP BY c.id, c.name
            ORDER BY student_count DESC LIMIT 5
            """, new { tenantId });

        // ---- Bài tập / Bài thi chờ chấm ----
        var pendingHomework = await _db.ScalarAsync<long>("""
            SELECT COUNT(*) FROM homework_submissions hs
            JOIN homework_assignment_students has2 ON has2.id=hs.assignment_student_id
            JOIN homework_assignments ha ON ha.id=has2.assignment_id
            WHERE ha.tenant_id=@tenantId AND hs.status='submitted'
            """, new { tenantId });

        var pendingMockExams = await _db.ScalarAsync<long>("""
            SELECT COUNT(*) FROM mock_exam_students mes
            JOIN mock_exams me ON me.id=mes.mock_exam_id
            WHERE me.tenant_id=@tenantId AND mes.status='submitted'
            """, new { tenantId });

        // ---- Build response (flat, khớp Frontend contract) ----
        return new
        {
            // Thẻ tổng quan
            students,
            activeClasses,
            teachers,
            todaySessions,
            pendingHomeworkGrading = pendingHomework,
            pendingMockExamGrading = pendingMockExams,

            // Tài chính
            revenue = new
            {
                byMethod = revenueByMethod,
                byDay = revenueByDay,
                total = totalRevenue,
                month = m,
                year = y,
            },
            expenses = new { total = totalExpenses },
            netProfit = totalRevenue - totalExpenses,
            transactions = totalTransactions,

            // Hóa đơn chờ
            pendingInvoices = new
            {
                count = Convert.ToInt64(pendingInvoice?.GetValueOrDefault("count") ?? 0),
                total = Convert.ToDouble(pendingInvoice?.GetValueOrDefault("total") ?? 0),
            },

            // Công nợ
            debtAging = debtAgingRaw == null ? new Dictionary<string, object>() : new Dictionary<string, object>(debtAgingRaw)
            {
                ["topDebtors"] = topDebtors,
            },

            // Chi nhánh
            branchStats,

            // Bổ trợ
            weeklyAttendance = weeklyAttendance.Select(r => new
            {
                date = r.GetValueOrDefault("session_date"),
                present = Convert.ToInt64(r.GetValueOrDefault("present_count") ?? 0),
                total = Convert.ToInt64(r.GetValueOrDefault("total_count") ?? 0),
                rate = Convert.ToInt64(r.GetValueOrDefault("total_count") ?? 0) > 0
                    ? (int)Math.Round(Convert.ToDouble(r.GetValueOrDefault("present_count") ?? 0)
                       / Convert.ToDouble(r.GetValueOrDefault("total_count") ?? 1) * 100)
                    : 0,
            }),
            studentStatusDistribution = studentStatusDist.Select(r =>
            {
                var s = r.GetValueOrDefault("study_status")?.ToString() ?? "";
                var labelMap = new Dictionary<string, string>
                {
                    ["active"]    = "Đang học",
                    ["paused"]    = "Tạm nghỉ",
                    ["graduated"] = "Đã tốt nghiệp",
                    ["dropped"]   = "Đã nghỉ",
                };
                return new
                {
                    status = (object?)s,
                    label  = (object?)(labelMap.TryGetValue(s, out var lbl) ? lbl : s),
                    count  = Convert.ToInt64(r.GetValueOrDefault("count") ?? 0),
                };
            }),
            topClasses = topClasses.Select(r => new
            {
                name = r.GetValueOrDefault("name"),
                count = Convert.ToInt64(r.GetValueOrDefault("student_count") ?? 0),
            }),
        };
    }
}
